### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Sites): functors which preserves sheafification

### PR DESCRIPTION
**Pull request by joelriou**
Given a Grothendieck topology `J` on `C` and `F : A ⥤ B`, we define a type class `J.PreservesSheafification F`. We say that `F` preserves the sheafification if whenever a morphism of presheaves `P₁ ⟶ P₂` induces an isomorphism on the associated sheaves, then the induced map `P₁ ⋙ F ⟶ P₂ ⋙ F` also induces an isomorphism on the associated sheaves.

If we also assume `J.HasSheafCompose F` (which says that the postcomposition with `F` induces a functor `sheafCompose J F : Sheaf J A ⥤ Sheaf J B`), then we obtain that the canonical map is an isomorphism `sheafify J (P ⋙ F) ≅ sheafify J P ⋙ F`.

We obtain `PreservesSheafification J (forget D)` when `D` is a concrete category satisfying suitable conditions.

---

- [x] depends on: #12374

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/PreservesSheafification.lean [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+namespace GrothendieckTopology
+
+variable {D : Type*} [Category.{max v u} D]
+  [ConcreteCategory.{max v u} D] [PreservesLimits (forget D)]
+  [∀ X : C, HasColimitsOfShape (J.Cover X)ᵒᵖ D]
+  [∀ X : C, PreservesColimitsOfShape (J.Cover X)ᵒᵖ (forget D)]
+  [∀ (α β : Type max u v) (fst snd : β → α),
+      Limits.HasLimitsOfShape (Limits.WalkingMulticospan fst snd) D]
+  [(forget D).ReflectsIsomorphisms]
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Sites/PreservesSheafification.lean:132]: Is it possible to prove this for general functors other than forget (or maybe at least forget_2) for this?

Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/PreservesSheafification.lean]: Yes. Actually, I think I should generalize `PreservesSheafification` without even assuming `HasSheafCompose`. For example, in some applications, we would like to apply this to the free abelian group functor from sets to abelian groups (or similarly the free module functor). The postcomposition with this free abelian group functor should not preserve sheaves, but still it behaves quite well with respect to sheafification, in the sense that if we introduce the class `W` of morphisms of presheaves which become isos after sheafification, then `W` is "preserved" by the application of this free abelian group functor. If we assume moreover `HasSheafCompose`, then this condition should be equivalent to the current draft `PreservesSheafification`. (Eventually, for good enough concrete categories, the class `W` should coincide with the class of morphisms of presheaves that are both locally injective and locally surjective.)

Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/PreservesSheafification.lean]: The class `W` is defined in #12374

Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/PreservesSheafification.lean]: I have generalized the definition of `PreservesSheafification` as I suggested above.

**Comment by erdOne** [comment]
Thanks a lot!
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**